### PR TITLE
[WIP] Release code of MixFormer (CVPR2022, Oral)

### PR DESCRIPTION
**Pull request by chensnathan**
[MixFormer: Mixing Features across Windows and Dimensions](https://arxiv.org/abs/2204.02557)

Pre-trained models will be added in next few days.

### CODE REVIEW

**Comment by Seperendity** [comment]
您好，想请教一下，ppcls/arch/backbone/model_zoo/mixformer.py中line229这里的维度`v = v * x_cnn2v`是如何计算的呢？我看每个部分的最后两个维度分别是`（1, C // self.num_heads）`和`(N, C // self.num_heads)`这里做矩阵乘列和行的维度不是不对应吗？

**Comment by chensnathan** [comment]
@Seperendity 你好，这个是可以通过[广播机制](https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/beginner/tensor_cn.html#id7)来实现的

**Comment by Seperendity** [comment]
@chensnathan 非常感谢您的解答！知道用的是广播机制了。但还是对为什么广播后的值乘的对应维度是`numswindow`和`tokens`数这两维，我看论文的意思以为是把权重乘到通道维度上。`x_cnn2v = torch.sigmoid(channel_interaction).reshape([-1, 1, self.num_heads, 1, C // self.num_heads])`和` v = v.reshape([x_cnn2v.shape[0], -1, self.num_heads, N, C // self.num_heads])`代码中这么乘的原因是什么呢？直观上来看并没有将学到的权重赋到dims维度上，希望您能解答一下，不甚感激。

**Comment by chensnathan** [comment]
@Seperendity 你好，这样做是为了配合v的维度。举个例子理解一下，假设v的shape是`[B, C, H, W]`，x_cnn2v的shape是`[B, C, 1, 1]`，那么`v = v * x_cnn2v`是一个简单的channel attention。但是，在代码里的第223行，因为后续要准备做window-based self-attention，v的shape是`[B*(H/win)*(W/win), win*win, num_heads, C/num_heads]`，而x_cnn2v的shape是`[B, C, 1, 1]`，这个时候没法直接做channel attention。当然这里可以用不同的实现：
1. 你可以把v再reshape回`[B, C, H, W]`，做完channel attention之后，再变成`[B*(H/win)*(W/win), win*win, num_heads, C/num_heads]`，再进入到下面的self-attention。
2. 我这里选择的是，v的shape从`[B*(H/win)*(W/win), win*win, num_heads, C/num_heads]`变为`[B, (H/win)*(W/win), win*win, num_heads, C/num_heads]`，x_cnn2v变为`[B, 1, 1, num_heads, C/num_heads]`，然后再变为`[B*(H/win)*(W/win), win*win, num_heads, C/num_heads]`，再进入到下面的self-attention。
本质上是一样的。

**Comment by Seperendity** [comment]
@chensnathan 明白您的意思了，非常感谢您的耐心解答！很有意思的工作

**Comment by cxz1276316542** [comment]
> [MixFormer: Mixing Features across Windows and Dimensions](https://arxiv.org/abs/2204.02557)
> 
> Pre-trained models will be added in next few day
你好，预训练模型出来了吗？在哪里下载呢？

File: docs/en/models/MixFormer_en.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+<a name='2'></a>
+## 2. Main Results
+
+<a name='2.1'></a>
+### Results on ImageNet-1K
+We provide the accuracy with FLOPs of MixFormer on ImageNet-1K. Unlike other vision transformer variants, they only show promising results when the FLOPs are pretty large (e.g., 4.5G). MixFormer can achieve favorable results **with small model sizes (FLOPs < 1G)**, which are nontrivial results.
+
+| Models | Top1 | Top5 | Reference top1| FLOPs (G) |
+|:--:|:--:|:--:|:--:|:--:|
```
Comment (cuicheng01) [inline docs/en/models/MixFormer_en.md]: 这个表格也需要改一下

File: docs/zh_CN/models/ImageNet1k/MixFormer.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+## 1. 介绍
+
+MixFormer是一个高效、通用的骨干网路（Vision Transformer）。在MixFormer中，主要有两个创新的设计：（1）通过平行分支的设计，将局部窗口自注意力（local-window self-attention）与Depthwise卷积进行组合，解决局部窗口自注意力的感受野受限的问题，（2）在平行分支之间提出双向交互模块，使得两个分支可以在channel和spatial两个维度都能实现信息互补，增强整体的建模能力。上述两个设计的组合使得MixFormer可以融合不同局部窗口与不同维度的特征信息，从而在图像分类以及5个重要的下游任务中取得比其他vision transformer骨干网络更好的结果。
+
+> [**MixFormer: Mixing Features across Windows and Dimensions**](https://arxiv.org/abs/2204.02557)<br> 
+> Qiang Chen, Qiman Wu, Jian Wang, Qinghao Hu, Tao Hu, Errui Ding, Jian Cheng, Jingdong Wang<br>
+> CVPR2022, **Oral** presentation
+
+![image](../../images/MixFormer/MixingBlock.png)
+
```
Comment (cuicheng01) [inline docs/zh_CN/models/ImageNet1k/MixFormer.md:24]: 图片不显示

File: ppcls/arch/backbone/__init__.py [hunk 3]
```diff
@@ -65,6 +65,7 @@
 from .model_zoo.cspnet import CSPDarkNet53
 from .model_zoo.pvt_v2 import PVT_V2_B0, PVT_V2_B1, PVT_V2_B2_Linear, PVT_V2_B2, PVT_V2_B3, PVT_V2_B4, PVT_V2_B5
 from .model_zoo.mobilevit import MobileViT_XXS, MobileViT_XS, MobileViT_S
+from ppcls.arch.backbone.model_zoo.mixformer import MixFormer_B0, MixFormer_B1, MixFormer_B2, MixFormer_B3, MixFormer_B4, MixFormer_B5, MixFormer_B6
```
Comment (cuicheng01) [inline ppcls/arch/backbone/__init__.py]: 这里参考其他模型改成一致的结构
